### PR TITLE
WT-7007 Backup architecture guide

### PR DESCRIPTION
**Pull request by kommiharibabu**
This architecture guide explain the details of how WT backup
is supported with internals.

### CODE REVIEW

File: src/docs/arch-backup.dox [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+As part of the process of collecting the list of files to be backed up, WT takes the
+schema lock to block any modifications to create/drop new tables in the database.
+In addition to table creation, there is a contract that all files that exist at the
+time the backup starts exist for the duration of the backup.
+
+As part of backup, WT internally collects all the log files also that are present in the
+database and stores them in a list to let them to be backed up. These log files are
+required to make the backup consistent. As part of collecting the log files, forces a
+log file switch and it does not allow any log files to be archived/removed. WT doesn't
+use any pre allocated log files and create new set of log files.
```
Comment (sueloverso) [inline src/docs/arch-backup.dox:26]: This paragraph feels a bit choppy. Maybe something like:
WiredTiger log files are also part of the generated file list for backup. There is a contract that the log files do not contain or make visible operations that are added to the log after the backup started that could have adverse affects during recovery and restore. Therefore WiredTiger forces a log file switch when opening the backup cursor. To fulfill the earlier contract that all files must exist during backup, WiredTiger does not use any pre-allocated log files nor does it archive or remove old log files during the time the backup cursor is open.

File: src/docs/arch-backup.dox [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+Once the backup is started successfully, any checkpoints that exist on the database
+before backup starts are retained until the backup finishes. All the newer checkpoints
```
Comment (sueloverso) [inline src/docs/arch-backup.dox:29]: "Once the backup is started" -> "Once the backup cursor is opened"
"are retained" -> "must be retained"
"until the backup finishes" -> "until the backup cursor is closed."

File: src/docs/arch-backup.dox [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+that are created during the backup in progress are cleaned whenever a newer checkpoint
+is created. So longer backup time should not increase the disk size of the database other
+than the expected growth.
```
Comment (sueloverso) [inline src/docs/arch-backup.dox:32]: While the last sentence refers to problems that existed in past years, it feels out of place now to a new reader. I suggest removing it.

File: src/docs/arch-backup.dox [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+Note: No two backups can run in parallel.
```
Comment (sueloverso) [inline src/docs/arch-backup.dox:34]: Note: There can only be one backup cursor open at a time other unless using duplicate backup cursors. <Maybe add a section on duplicate backup cursors, when/why they're used and refer to it here.>

File: src/docs/arch-backup.dox [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+# Block based incremental backup #
+
+Block based incremental backup is performed by tracking the modified blocks in the checkpoint.
+whenever a new checkpoint occurs on a table, the new blocks that are allocated for this
```
Comment (sueloverso) [inline src/docs/arch-backup.dox:39]: Capital W for "Whenever..."
"... the new blocks that are allocated..." is insufficient because existing blocks get dirty too. I suggest:
 "...on a table, any new or modified blocks in this checkpoint are..."

File: src/docs/arch-backup.dox [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+checkpoint are recorded as a bit string. This bit string information is gets updated with
```
Comment (sueloverso) [inline src/docs/arch-backup.dox:40]: "is gets" remove "gets"

File: src/docs/arch-backup.dox [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+every new checkpoint.
+
+Whenever the incremental backup cursor is opened to perform the backup. WT returns all the
+modified file offsets and sizes that are modified after the full backup. WT returns full file
```
Comment (sueloverso) [inline src/docs/arch-backup.dox:44]: This sentence has 'modified' twice. Remove the first one.
"...returns the full file information to be copied"
I also suggest adding a reference to @ref backup_incremental-block

File: src/docs/arch-backup.dox [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+to be copied for all the new files that are created, renamed or imported into the database
+after incremental backup is started.
+
+# Log based incremental backup #
+
+When backup cursor is opened with target as \c log: the log based incremental backup
```
Comment (sueloverso) [inline src/docs/arch-backup.dox:50]: "When the backup cursor is opened with the \c target..."
Add a reference to @ref backup_incremental

File: src/docs/arch-backup.dox [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+is performed by adding all the existing log files in the database to the list of files
+that needs to be copied.
```
Comment (sueloverso) [inline src/docs/arch-backup.dox:52]: There is more to log based incremental. Log archiving must be turned off. The user can archive no longer needed log files instead by calling truncate (See the backup doc for information and details.) Please include some of that information here.

File: src/docs/arch-backup.dox [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+# Target backup #
+
+It is a kind of full backup but WT collects only the target tables that are specified
```
Comment (sueloverso) [inline src/docs/arch-backup.dox:56]: Please expand on an ambiguous "It".

**Review by sueloverso — COMMENTED** [review]
This is a good start. I have a bunch of edits and suggestions.

File: src/docs/arch-backup.dox [hunk 11]
```diff
@@ -0,0 +1,60 @@
+/*! @arch_page arch-backup Backup
+
+# Overview #
+
+Backup in WiredTiger can be performed using the \c "backup:" cursors. WiredTiger
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:5]: Fixed.

File: src/docs/arch-backup.dox [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+supports four types of backups.
+
+1. Full database
+2. Block based incremental backup
+3. Log based incremental backup
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:10]: Fixed.

File: src/docs/arch-backup.dox [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+4. Target backup
+
+# Full database #
+
+When the application opens the backup cursor, internally WT collects all the tables
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:15]: Fixed as you suggested.

File: src/docs/arch-backup.dox [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+that are present in the database and stores them in a list to let them be backed up.
+As part of the process of collecting the list of files to be backed up, WT takes the
+schema lock to block any modifications to create/drop new tables in the database.
+In addition to table creation, there is a contract that all files that exist at the
+time the backup starts exist for the duration of the backup.
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:20]: Thanks, Fixed.

File: src/docs/arch-backup.dox [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+2. Block based incremental backup
+3. Log based incremental backup
+4. Target backup
+
+# Full database #
+
+When the application opens the backup cursor, internally WT collects all the tables
+that are present in the database and stores them in a list to let them be backed up.
+As part of the process of collecting the list of files to be backed up, WT takes the
+schema lock to block any modifications to create/drop new tables in the database.
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:18]: Fixed.

File: src/docs/arch-backup.dox [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+# Overview #
+
+Backup in WiredTiger can be performed using the \c "backup:" cursors. WiredTiger
+supports four types of backups.
+
+1. Full database
+2. Block based incremental backup
+3. Log based incremental backup
+4. Target backup
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:11]: Fixed.

File: src/docs/arch-backup.dox (hunk 1 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:26]: Thanks, fixed as suggested.

File: src/docs/arch-backup.dox (hunk 2 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:29]: Fixed.

File: src/docs/arch-backup.dox (hunk 3 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:32]: Removed.

File: src/docs/arch-backup.dox (hunk 4 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:34]: Added as suggested.

File: src/docs/arch-backup.dox (hunk 5 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:39]: Fixed.

File: src/docs/arch-backup.dox (hunk 6 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:40]: Fixed.

File: src/docs/arch-backup.dox (hunk 7 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:44]: Fixed.

File: src/docs/arch-backup.dox (hunk 10 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:56]: Expanded.

File: src/docs/arch-backup.dox (hunk 9 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:52]: Added the details of log archiving.

File: src/docs/arch-backup.dox (hunk 8 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:50]: Fixed.

File: src/docs/arch-backup.dox [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+Backup in WiredTiger can be performed using the \c "backup:" cursor. WiredTiger
+supports four types of backups.
+
+1. Full database
+2. Block-based incremental backup
+3. Log-based incremental backup
+4. Target backup
+
+The following page describes how backup works inside the WiredTiger. Please refer @ref backup
```
Comment (sueloverso) [inline src/docs/arch-backup.dox:13]: 'Please refer to @ref..."

File: src/docs/arch-backup.dox [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+created.
+
+# Block-based incremental backup #
+
+Block-based incremental backup is performed by tracking the modified blocks in the checkpoint.
+Whenever a new checkpoint occurs on a file, any new or modified blocks in this checkpoint are
+recorded as a bit string. This bit string information is updated with every new checkpoint.
+
+Whenever the incremental backup cursor is opened to perform the backup. WiredTiger returns all
+the file offsets and sizes that are modified after the full backup. WiredTiger returns the full
```
Comment (sueloverso) [inline src/docs/arch-backup.dox:44]: 'after the full backup' is not quite right because the modified blocks could be from the full backup or simply an earlier (i.e. the previous) incremental backup. I suggest:
WiredTiger returns all the file offsets and sizes that are modified from the previous source backup identifier.

File: src/docs/arch-backup.dox [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+more information on how to use the block-based incremental backup.
+
+# Log-based incremental backup #
+
+When backup cursor is opened with target as \c "target=(\"log:\\")" the log-based incremental
+backup is performed by adding all the existing log files in the database to the list of files
+that needs to be copied. Applications wanting to use log files for incremental backup must
+first disable automatic log file removal using the \c log=(archive=false) configuration to
+::wiredtiger_open. By default, WiredTiger automatically removes log files no longer required
+for recovery. Refer @ref backup_incremental for more information on how to use the log-based
```
Comment (sueloverso) [inline src/docs/arch-backup.dox:56]: 'Refer to @ref`

File: src/docs/arch-backup.dox [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+incremental backup.
+
+# Target backup #
+
+When backup cursor is opened with target as \c "target=" WiredTiger internally generates list
```
Comment (sueloverso) [inline src/docs/arch-backup.dox:61]: 'When the backup cursor..'
'...is opened with the \c target configuration string as...'

File: src/docs/arch-backup.dox [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+of targetted files instead of all files like the full backup. If the targetted object is a
+table, all the indexes and column groups of the table are also backed up.
+
+Note: There can only be one backup cursor open at a time other unless using duplicate backup
+cursors. The duplicate backup cursors are used for catching up with the log files or with
+block-based incremental backup. Please refer @ref backup_duplicate for more details.
```
Comment (sueloverso) [inline src/docs/arch-backup.dox:67]: 'Please refer to @ref...'

File: src/docs/arch-index.dox [hunk 22]
```diff
@@ -127,6 +127,10 @@ We go into some detail for some of the internal components.
 
 The Block Manager manages the reading and writing of disk blocks.
 
+@subpage arch-backup
+
+The backup is a type of cursor used to backup the database.
```
Comment (sueloverso) [inline src/docs/arch-index.dox:132]: I'd reword as:
'Hot backup uses a type of cursor to backup the database.'

**Review by sueloverso — APPROVED** [review]
This is very close. I have several minor wording changes.
I am LGTM unless there is a substantial rewriting of anything based on any other review.

File: src/docs/arch-backup.dox [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+When the application opens the backup cursor, internally WT collects all the tables
+that are present in the database and stores them in a list to let them be backed up.
+As part of the process of collecting the list of files to be backed up, WT takes the
+schema lock to block any modifications to create/drop new tables in the database.
+In addition to table creation, there is a contract that all files that exist at the
+time the backup starts exist for the duration of the backup.
+
+As part of backup, WT internally collects all the log files also that are present in the
+database and stores them in a list to let them to be backed up. These log files are
+required to make the backup consistent. As part of collecting the log files, forces a
```
Comment (ddanderson) [inline src/docs/arch-backup.dox]: ...As part of collecting the log files, _WiredTiger_ forces

File: src/docs/arch-backup.dox [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+log file switch and it does not allow any log files to be archived/removed. WT doesn't
```
Comment (ddanderson) [inline src/docs/arch-backup.dox]: Use `WiredTiger` instead of `WT` (throughout the file).

File: src/docs/arch-backup.dox (hunk 1 above)
Comment (ddanderson) [inline src/docs/arch-backup.dox]: This sentence feels unclear to me.  `doesn't use pre allocated log files`. When?  for the duration of the backup?  `and creates a new set of log files`.  Doesn't it create just one new one?  And when?  At the beginning of the backup?  Could be two sentences to clarify.

File: src/docs/arch-backup.dox (hunk 2 above)
Comment (ddanderson) [inline src/docs/arch-backup.dox]: ...before _the backup started_ are....

File: src/docs/arch-backup.dox [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+use any pre allocated log files and create new set of log files.
+
+Once the backup is started successfully, any checkpoints that exist on the database
+before backup starts are retained until the backup finishes. All the newer checkpoints
+that are created during the backup in progress are cleaned whenever a newer checkpoint
+is created. So longer backup time should not increase the disk size of the database other
```
Comment (ddanderson) [inline src/docs/arch-backup.dox]: ...So _a_ longer backup time...   or  ...longer time spent in backup...

File: src/docs/arch-backup.dox (hunk 5 above)
Comment (ddanderson) [inline src/docs/arch-backup.dox]: Capitalize `whenever`.

File: src/docs/arch-backup.dox (hunk 6 above)
Comment (ddanderson) [inline src/docs/arch-backup.dox]: `is gets`....  pick one :-)

File: src/docs/arch-backup.dox [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+
+Block based incremental backup is performed by tracking the modified blocks in the checkpoint.
+whenever a new checkpoint occurs on a table, the new blocks that are allocated for this
+checkpoint are recorded as a bit string. This bit string information is gets updated with
+every new checkpoint.
+
+Whenever the incremental backup cursor is opened to perform the backup. WT returns all the
+modified file offsets and sizes that are modified after the full backup. WT returns full file
+to be copied for all the new files that are created, renamed or imported into the database
+after incremental backup is started.
```
Comment (ddanderson) [inline src/docs/arch-backup.dox]: after _the_ incremental.

File: src/docs/arch-backup.dox (hunk 10 above)
Comment (ddanderson) [inline src/docs/arch-backup.dox]: Avoid starting a paragraph with 'it'.

**Review by ddanderson — APPROVED** [review]
@kommiharibabu sorry for my slow review.  This looks pretty good to me (tho @sueloverso is the expert on the content), my comments are minor.

**Comment by ddanderson** [comment]
@kommiharibabu please _ignore_ my last review, I had an outdated version open and it looks like you've fixed most of my comments.  I will rereview.

File: src/docs/arch-backup.dox [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+for details on how to use each of these types of backups.
+
+# Full database #
+
+When the application opens the backup cursor, internally WiredTiger generates a list of all
+files in the database that are necessary for the backup. When generating that list WiredTiger
+takes both the checkpoint and schema locks to block database file modifications. There is a
+contract that all files that exist at the time the backup starts exist for the entire time
+the backup cursor is open. This contract applies to all files in the database, not just files
+that are included in the generated list.
```
Comment (ddanderson) [inline src/docs/arch-backup.dox:23]: One thing that isn't clear is how long the checkpoint/schema locks are held - is it while generating the list or for the length of the backup?

File: src/docs/arch-backup.dox [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+that are created during the backup in progress are cleaned whenever a newer checkpoint is
+created.
+
+# Block-based incremental backup #
+
+Block-based incremental backup is performed by tracking the modified blocks in the checkpoint.
+Whenever a new checkpoint occurs on a file, any new or modified blocks in this checkpoint are
+recorded as a bit string. This bit string information is updated with every new checkpoint.
+
+Whenever the incremental backup cursor is opened to perform the backup. WiredTiger returns all
```
Comment (ddanderson) [inline src/docs/arch-backup.dox:43]: Should be a comma after "perform the backup".

File: src/docs/arch-backup.dox (hunk 20 above)
Comment (ddanderson) [inline src/docs/arch-backup.dox:61]: ....internally generates _a_ list...

File: src/docs/arch-backup.dox [hunk 29]
```diff
@@ -0,0 +1,69 @@
...
+incremental backup.
+
+# Target backup #
+
+When backup cursor is opened with target as \c "target=" WiredTiger internally generates list
+of targetted files instead of all files like the full backup. If the targetted object is a
+table, all the indexes and column groups of the table are also backed up.
+
+Note: There can only be one backup cursor open at a time other unless using duplicate backup
+cursors. The duplicate backup cursors are used for catching up with the log files or with
```
Comment (ddanderson) [inline src/docs/arch-backup.dox:66]: Is it allowed to have multiple duplicate cursors open, to allow parallelism with incremental backup?

**Review by ddanderson — APPROVED** [review]
Looking pretty good.  I had a couple small comments and a couple questions that should be clarified in the text.

File: src/docs/arch-backup.dox (hunk 17 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:13]: Fixed.

File: src/docs/arch-backup.dox (hunk 28 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:43]: Fixed.

File: src/docs/arch-backup.dox [hunk 30]
```diff
@@ -0,0 +1,69 @@
...
+# Block-based incremental backup #
+
+Block-based incremental backup is performed by tracking the modified blocks in the checkpoint.
+Whenever a new checkpoint occurs on a file, any new or modified blocks in this checkpoint are
+recorded as a bit string. This bit string information is updated with every new checkpoint.
+
+Whenever the incremental backup cursor is opened to perform the backup. WiredTiger returns all
+the file offsets and sizes that are modified after the full backup. WiredTiger returns the full
+file information to be copied for all the new files that are created, renamed or imported into
+the database after incremental backup cursor is opened. Refer @ref backup_incremental-block for
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:46]: fixed

File: src/docs/arch-backup.dox [hunk 31]
```diff
@@ -0,0 +1,69 @@
...
+more information on how to use the block-based incremental backup.
+
+# Log-based incremental backup #
+
+When backup cursor is opened with target as \c "target=(\"log:\\")" the log-based incremental
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:51]: Fixed.

File: src/docs/arch-backup.dox (hunk 19 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:56]: Fixed.

File: src/docs/arch-backup.dox [hunk 32]
```diff
@@ -0,0 +1,69 @@
...
+for recovery. Refer @ref backup_incremental for more information on how to use the log-based
+incremental backup.
+
+# Target backup #
+
+When backup cursor is opened with target as \c "target=" WiredTiger internally generates list
+of targetted files instead of all files like the full backup. If the targetted object is a
+table, all the indexes and column groups of the table are also backed up.
+
+Note: There can only be one backup cursor open at a time other unless using duplicate backup
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:65]: Fixed.

File: src/docs/arch-backup.dox (hunk 21 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:67]: Fixed.

File: src/docs/arch-backup.dox (hunk 20 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:61]: fixed.

Comment (kommiharibabu) [reply src/docs/arch-backup.dox:61]: fixed.

File: src/docs/arch-backup.dox (hunk 27 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:23]: I reordered the sentence to give clarity on how long the locks are held.

File: src/docs/arch-backup.dox (hunk 18 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:44]: Fixed as suggested.

File: src/docs/arch-backup.dox (hunk 29 above)
Comment (kommiharibabu) [reply src/docs/arch-backup.dox:66]: AFAIK Yes, multiple duplicate cursors are possible to open to allow parallelism.

File: src/docs/arch-index.dox (hunk 22 above)
Comment (kommiharibabu) [reply src/docs/arch-index.dox:132]: fixed as suggested.
